### PR TITLE
[SPARK-44963][PYTHON][ML][TESTS] Make PySpark (pyspark-ml module) tests passing without any optional dependency

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

This PR proposes to fix the tests to properly run or skip when there aren't optional dependencies installed.

### Why are the changes needed?

Currently, it fails as below:

```
./python/run-tests --python-executables=python3 --modules=pyspark-ml
...
Starting test(python3): pyspark.ml.tests.test_model_cache (temp output: /Users/hyukjin.kwon/workspace/forked/spark/python/target/f6f88c1e-0cb2-43e6-980e-47f1cdb9b463/python3__pyspark.ml.tests.test_model_cache__zij05l1u.log)
Traceback (most recent call last):
  File "/Users/hyukjin.kwon/miniconda3/envs/vanilla-3.10/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/Users/hyukjin.kwon/miniconda3/envs/vanilla-3.10/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/Users/hyukjin.kwon/workspace/forked/spark/python/pyspark/ml/tests/test_functions.py", line 18, in <module>
    import pandas as pd
ModuleNotFoundError: No module named 'pandas'
```

PySpark tests should pass without optional dependencies.

### Does this PR introduce _any_ user-facing change?

No, test-only.

### How was this patch tested?

Manually ran as described above.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

File: python/pyspark/ml/torch/tests/test_distributor.py [hunk 1]
```diff
@@ -18,7 +18,7 @@
 import contextlib
 import os
 import shutil
-from six import StringIO
```
Comment (HyukjinKwon) [inline python/pyspark/ml/torch/tests/test_distributor.py:21]: we don't need backward compat for Python 2

**Review by dongjoon-hyun — DISMISSED** [review]
+1, LGTM (Pending CIs). Thank you so much for this, @HyukjinKwon .

**Comment by dongjoon-hyun** [comment]
It seems that `python-ml` UT fails. Could you re-trigger the failed test pipeline to make it sure?

**Review by dongjoon-hyun — COMMENTED** [review]
BTW, it seems that we need `numpy` still.

```
$ ./python/run-tests --python-executables=python3 --modules=pyspark-ml
Running PySpark tests. Output is in /Users/dongjoon/APACHE/spark-merge/python/unit-tests.log
Will test against the following Python executables: ['python3']
Will test the following Python modules: ['pyspark-ml']
python3 python_implementation is CPython
python3 version is: Python 3.11.4
Starting test(python3): pyspark.ml.tests.connect.test_legacy_mode_feature (temp output: /Users/dongjoon/APACHE/spark-merge/python/target/333d370f-3a76-48b3-967a-c16cc542f5c1/python3__pyspark.ml.tests.connect.test_legacy_mode_feature__lo19yzgf.log)
Starting test(python3): pyspark.ml.tests.connect.test_legacy_mode_classification (temp output: /Users/dongjoon/APACHE/spark-merge/python/target/29922d0f-5b14-473c-bfad-eae42fb2779d/python3__pyspark.ml.tests.connect.test_legacy_mode_classification__sgsncqxj.log)
Starting test(python3): pyspark.ml.tests.connect.test_legacy_mode_evaluation (temp output: /Users/dongjoon/APACHE/spark-merge/python/target/0fab9221-842f-4569-afd1-db03494469fa/python3__pyspark.ml.tests.connect.test_legacy_mode_evaluation__5008t5g_.log)
Starting test(python3): pyspark.ml.tests.connect.test_legacy_mode_pipeline (temp output: /Users/dongjoon/APACHE/spark-merge/python/target/36fcb2b9-2d47-4a96-bb85-deec88d74e1d/python3__pyspark.ml.tests.connect.test_legacy_mode_pipeline__kgqnp3gs.log)
Traceback (most recent call last):
  File "<frozen runpy>", line 189, in _run_module_as_main
  File "<frozen runpy>", line 112, in _get_module_details
  File "/Users/dongjoon/APACHE/spark-merge/python/pyspark/ml/__init__.py", line 22, in <module>
    from pyspark.ml.base import (
  File "/Users/dongjoon/APACHE/spark-merge/python/pyspark/ml/base.py", line 40, in <module>
    from pyspark.ml.param import P
  File "/Users/dongjoon/APACHE/spark-merge/python/pyspark/ml/param/__init__.py", line 32, in <module>
    import numpy as np
ModuleNotFoundError: No module named 'numpy'
```

**Comment by dongjoon-hyun** [comment]
I added `optional` into the PR title because `numpy` is required dependency for MLLib.
- https://spark.apache.org/docs/latest/api/python/getting_started/install.html#dependencies

**Comment by HyukjinKwon** [comment]
Merged to master.

**Review by dongjoon-hyun — COMMENTED** [review]
+1, LGTM.

**Pull request closed without merging**